obsplus/structures/grid.py

```diff
@@ -0,0 +1,1323 @@
+        """
+    # For each layer in vel_input, apply the specified velocity
+    for elev in layers:
+        if not isinstance(elev, Iterable) or isinstance(elev, str):
+            raise TypeError(
+                "velocity input should be a 2D list of values and elevations"
+            )
+        grid.values[:, :, grid.grid_points[2] <= elev[1]] = elev[0]
+
+
+def apply_rectangles(
+    grid, rectangles, tol=1e-6, conversion=None, conversion_kwargs=None
+):
+    """
+    Internal method for perturbing grid values in rectangular grid regions
+
+    Parameters
+    ----------
+    grid : Grid
+        Grid on which to apply the layers
+    rectangles : pandas DataFrame (required)
+        List of velocity perturbations to apply. Required columns
+        include ["DELTA", "XMIN", "XMAX", "YMIN", "YMAX"] where "DELTA" is
+        some percentage of the current grid value
+    tol : float (default=1e-6)
+        Value to add to rectangle coordinates to prevent rounding
+        errors. This value should be small relative to the grid
+        spacing.
+    conversion : list-like (default=None)
+        See convert_coords
+    conversion_kwargs : dict (default=None)
+        See convert_coords
+    """
+    if isinstance(rectangles, str):
+        path = rectangles
+        if not os.path.isfile(path):
+            raise OSError(f"rectangles file does not exist: {path}")
+        rectangles = read_file(path)
+        cols = {"DELTA", "XMIN", "YMIN", "ZMIN", "XMAX", "YMAX", "ZMAX"}
+        if not cols.issubset(rectangles.columns):
+            raise IOError(f"{path} is not a valid rectangles file")
+    elif isinstance(rectangles, pd.DataFrame):
+        pass
+    else:
+        raise TypeError("rectangles must be a pandas DataFrame")
+    if conversion:
+        temp_points = pd.DataFrame(columns=["X", "Y", "Z"])
+        i = 0
+        for num, r in rectangles.iterrows():
+            temp_points.loc[i] = [r.XMIN, r.YMIN, r.ZMIN]
+            temp_points.loc[i + 1] = [r.XMAX, r.YMAX, r.ZMAX]
+            i = i + 2
+        temp_points = convert_coords(
+            temp_points,
+            conversion=conversion,
+            conversion_kwargs=conversion_kwargs,
+            xout="X",
+            yout="Y",
+            zout="Z",
+        )
+        for num, r in temp_points.iterrows():  # Why did I do this this way?
+            i = num // 2
+            ind = rectangles.iloc[i].name
+            if (num % 2) == 0:
+                rectangles.loc[ind, "XMIN"] = r.X
+                rectangles.loc[ind, "YMIN"] = r.Y
+                rectangles.loc[ind, "ZMIN"] = r.Z
+            else:
+                rectangles.loc[ind, "XMAX"] = r.X
+                rectangles.loc[ind, "YMAX"] = r.Y
+                rectangles.loc[ind, "ZMAX"] = r.Z
+    v = grid.values
+    gmap = grid.grid_map
+    for num, zone in rectangles.iterrows():
+        delta = 0.01 * zone.DELTA
+        # Use a ridiculous one-liner to modify the zones within the
```

Comment:
Looks like it turned into a 18 liner.

Reply:
Indeed it did... thank you, black. I really should see if I can find a way to clean that up.